Better error msg for autograd profiler + multi-worker dataloader crash

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31473 Better error msg for autograd profiler + multi-worker dataloader crash**

Mitigates #6313

A common use case for the autograd profiler is to use it to run over an
entire model, including dataloading. The following will crash:
- run autograd profiler in CUDA mode
- Use a multi-worker DataLoader (presumably with the 'fork' spawn
method)
- because the autograd profiler initializes CUDA and forking after CUDA is
initialized is bad.

This PR puts in a nice error message when this happens so that users
aren't too confused. The new error message looks like:
https://gist.github.com/zou3519/903f15c3e86bad4585b7e5ce14cc1b70

Test Plan:
- Tested locally.
- I didn't add a test case for this because it's hard to write a test
case that doesn't completely stop the rest of our test suite from
running.

Differential Revision: [D19178080](https://our.internmc.facebook.com/intern/diff/D19178080)

zou3519 committed:
Better error msg for autograd profiler + multi-worker dataloader crash

Mitigates #6313

A common use case for the autograd profiler is to use it to run over an
entire model, including dataloading. The following will crash:
- run autograd profiler in CUDA mode
- Use a multi-worker DataLoader (presumably with the 'fork' spawn
method)
- because the autograd profiler initializes CUDA and forking after CUDA is
initialized is bad.

This PR puts in a nice error message when this happens so that users
aren't too confused. The new error message looks like:
https://gist.github.com/zou3519/903f15c3e86bad4585b7e5ce14cc1b70

Test Plan:
- Tested locally.
- I didn't add a test case for this because it's hard to write a test
case that doesn't completely stop the rest of our test suite from
running.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 714e346:


* **1/1** failures introduced in this PR
* **1/1** recognized as flaky
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=714e3462dc79050dfc63ce41b7a73bb1d62f7a2d).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4017296) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77591309) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77591309))
<details>
<summary>
<code>E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/xenial/pool/stable/amd64/docker-ce_18.09.4~3-0~ubuntu-xenial_amd64.deb  GnuTLS recv error (-54): Error in the pull function.</code>
</summary>

```
                                                                   82% [Working]               Get:26 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 intel-microcode amd64 3.20191115.1ubuntu0.16.04.2 [2,408 kB] 
                                             84% [Working]               Get:27 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 amd64-microcode amd64 3.20191021.1+really3.20180524.1~ubuntu0.16.04.2 [30.8 kB] 
                                             85% [Working]               Get:28 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 linux-image-generic amd64 4.4.0.170.178 [2,446 B] 
85% [28 linux-image-generic 2,446 B/2,446 B 100%]                  18.5 MB/s 0s 85% [Working]                                                      18.5 MB/s 0s                                                                                 Get:29 http://archive.ubuntu.com/ubuntu xenial/universe amd64 moreutils amd64 0.57-1 [55.0 kB] 
85% [29 moreutils 16.5 kB/55.0 kB 30%]                             18.5 MB/s 0s 86% [Working]                                                      18.5 MB/s 0s                                                                                 Get:30 http://archive.ubuntu.com/ubuntu xenial/main amd64 tcl8.6 amd64 8.6.5+dfsg-2 [14.2 kB] 
86% [30 tcl8.6 4,871 B/14.2 kB 34%]                                18.5 MB/s 0s 86% [Working]                                                      18.5 MB/s 0s                                                                                 Get:31 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 thermald amd64 1.5-2ubuntu4 [187 kB] 
86% [31 thermald 4,869 B/187 kB 3%]                                18.5 MB/s 0s 87% [Working]                                                      18.5 MB/s 0s                                                                                 Get:32 http://archive.ubuntu.com/ubuntu xenial/main amd64 tk8.6 amd64 8.6.5-1 [12.2 kB] 
87% [32 tk8.6 4,871 B/12.2 kB 40%]                                 18.5 MB/s 0s 88% [Working]                                                      18.5 MB/s 0s                                                                                 Get:33 http://archive.ubuntu.com/ubuntu xenial/main amd64 xterm amd64 322-1ubuntu1 [607 kB] 
88% [33 xterm 4,869 B/607 kB 1%]                                   18.5 MB/s 0s 89% [Working]                                                      18.5 MB/s 0s                                                                                 Fetched 113 MB in 7s (16.2 MB/s) 
W: --force-yes is deprecated, use one of the options starting with --allow instead. 
E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/xenial/pool/stable/amd64/docker-ce_18.09.4~3-0~ubuntu-xenial_amd64.deb  GnuTLS recv error (-54): Error in the pull function. 
 
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing? 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

anjali411 commented:
LGTM :) do you think we should also update the documentation to make it more clear that DataLoader is only supported for num_workers=0?

anjali411 reviewed (approved):
LGTM :) do you think we should also update the documentation to make it more clear that DataLoader is only supported for num_workers=0?

zou3519 commented:
> LGTM :) do you think we should also update the documentation to make it more clear that DataLoader is only supported for num_workers=0?

That's a good idea, I'll update the docs for https://pytorch.org/docs/stable/autograd.html#torch.autograd.profiler.profile as well in this PR

zou3519 committed:
Update on "Better error msg for autograd profiler + multi-worker dataloader crash"


Mitigates #6313

A common use case for the autograd profiler is to use it to run over an
entire model, including dataloading. The following will crash:
- run autograd profiler in CUDA mode
- Use a multi-worker DataLoader (presumably with the 'fork' spawn
method)
- because the autograd profiler initializes CUDA and forking after CUDA is
initialized is bad.

This PR puts in a nice error message when this happens so that users
aren't too confused. The new error message looks like:
https://gist.github.com/zou3519/903f15c3e86bad4585b7e5ce14cc1b70

Test Plan:
- Tested locally.
- I didn't add a test case for this because it's hard to write a test
case that doesn't completely stop the rest of our test suite from
running.

Differential Revision: [D19178080](https://our.internmc.facebook.com/intern/diff/D19178080)

[ghstack-poisoned]

zou3519 commented:
pytorch_python_doc_push (the only thing that matters for this PR) is green. All of the failed tests are unrelated, so I am shipping this.
